### PR TITLE
Move inbuilt Flavors into the Database

### PR DESCRIPTION
**Pull request by AlexejPenner**
## Describe changes
The single source of truth for flavors has moved into the database. The flavor registry has a method FlavorRegistry().register_flavors() responsible for uploading all flavors into the DB.

The sync happens in parallel with every alembic migration within the sql_zen_store.initialize() method.

This PR also adds two fields to flavors. A logo url and a docs url.

I have also created a convenience cli function to help during development. This function will help verify that all logos are valid. 

## Verify logo urls
`./zen-dev flavors verify logos`  # This will verify all logo urls

## Verify doc urls
This will verify all doc urls

`./zen-dev flavors verify docs`  

When there are changes on the docs, this can be simulated using out gitbook testpage

`./zen-dev flavors verify docs -r https://zenml-io.gitbook.io/test-space/`  

## Sync flavors
Using the following command, everyone developing a new flavor can manually trigger the sync to store the flavor in the db during local development
`./zen-dev flavors sync`

TODO:
[] Tests
[] Documentation 

## TODOs before this can be merged
[] Ensure Migration script does not break previously registered custom flavors

## Pre-requisites
Please ensure you have done the following:
- [ ] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: src/zenml/zen_stores/sql_zen_store.py
Comment (htahir1) [inline src/zenml/zen_stores/sql_zen_store.py:1438]: Maybe handle exceptions here more gracefully?

File: src/zenml/models/flavor_models.py
Comment (htahir1) [inline src/zenml/models/flavor_models.py:91]: Will this tinyurl expire at some point? It seems like it takes me to https://github.com/zenml-io. Is that the intended behavior?

File: src/zenml/enums.py
Comment (htahir1) [inline src/zenml/enums.py:156]: Dont we want a more generic name as @fa9r might reuse this for the metadata story?

**Review by htahir1 — CHANGES_REQUESTED** [review]
Left some comments but id ask @fa9r  and @stefannica to take a look at this as well to get some more opinions into this debate

File: src/zenml/models/flavor_models.py
Comment (AlexejPenner) [reply src/zenml/models/flavor_models.py:91]: will be replaced before merge once we decide on a default logo

File: src/zenml/enums.py
Comment (AlexejPenner) [reply src/zenml/enums.py:156]: I would argue, that the Metadata story will use very different types.

File: src/zenml/zen_stores/schemas/flavor_schemas.py
Comment (bcdurak) [inline src/zenml/zen_stores/schemas/flavor_schemas.py:71]: Can we perhaps make this an optional parameter?

File: src/zenml/zen_stores/sql_zen_store.py
Comment (bcdurak) [inline src/zenml/zen_stores/sql_zen_store.py:1430]: Just out of curiosity, is there a reason why we moved the `json.dumps` here? The `RequestModel` also has a dict instead of a str for the config schema.

**Review by bcdurak — COMMENTED** [review]
The changes look really good overall. I left a few small comments.

I may have missed the answer in the code, but I have one question here. Where does the registration of built-in and integrated flavors to our DB happen? Previously, these flavors were registered when you create an instance of the flavor registry. However, the imports are removed, and unless there is no migration from a different version, these flavors will not be registered to our DB,

File: src/zenml/zen_stores/sql_zen_store.py
Comment (AlexejPenner) [reply src/zenml/zen_stores/sql_zen_store.py:1430]: Having this a json in the FlavorResponseModel makes the serialization of that model very ugly, as you now have a string representation of a json within another json. This looks much cleaner and makes the dict immediatyl useable within python.

File: src/zenml/zen_stores/schemas/flavor_schemas.py
Comment (AlexejPenner) [reply src/zenml/zen_stores/schemas/flavor_schemas.py:71]: sure

**Comment by AlexejPenner** [comment]
> The changes look really good overall. I left a few small comments.
> 
> I may have missed the answer in the code, but I have one question here. Where does the registration of built-in and integrated flavors to our DB happen? Previously, these flavors were registered when you create an instance of the flavor registry. However, the imports are removed, and unless there is no migration from a different version, these flavors will not be registered to our DB,

This all happens as part of the migration. I generated a json file with all of our flavors, and those then get added to the DB

**Review by htahir1 — APPROVED** [review]

**Review by bcdurak — APPROVED** [review]

**Comment by AlexejPenner** [comment]
Thanks for the approvals, however just for stability i would not merge this into the pre-Xmas release as It its untested how this interacts with preexisting custom flavors.

**Comment by AlexejPenner** [comment]
Also remember to add the new sagemaker orchestrator to the migration json-file before merging this after the release.

**Comment by schustmi** [comment]
Just a general question from my side: Does that mean once this PR is merged we need to write migrations for each new flavor?

**Comment by AlexejPenner** [comment]
> Just a general question from my side: Does that mean once this PR is merged we need to write migrations for each new flavor?

Yes, it would mean that. Also all significant changes to current flavors (like path changes or naming changes) would need to be migrated (with appropriate downgrade) 

Additionally, the flavor registry might need to be deprecated somewhere down the line. 

I believe this might be a halfway point to eventually removing the in-built flavors out of the package and installing them as needed from separate packages, but I think that remains to be debated.

**Comment by schustmi** [comment]
> > Just a general question from my side: Does that mean once this PR is merged we need to write migrations for each new flavor?
> 
> Yes, it would mean that. Also all significant changes to current flavors (like path changes or naming changes) would need to be migrated (with appropriate downgrade)
> 
> Additionally, the flavor registry might need to be deprecated somewhere down the line.
> 
> I believe this might be a halfway point to eventually removing the in-built flavors out of the package and installing them as needed from separate packages, but I think that remains to be debated.

Thanks!

Good point regarding the removal, I see moving the flavors into separate packages as a necessity before going to 1.0 as otherwise breaking changes will occur much more often.
Regarding the flavor registry: Can't we already remove it in this PR? Or is there any use for it still?

**Review by schustmi — CHANGES_REQUESTED** [review]
Just as a reminder: If this get's merged without the automatic registry <-> db sync, we need to add the local and kaniko image builder in the migration script

File: src/zenml/artifact_stores/local_artifact_store.py
Comment (schustmi) [inline src/zenml/artifact_stores/local_artifact_store.py:159]: artifact_Store with capital `S`?

File: src/zenml/zen_stores/migrations/versions/bea8a6ce3015_port_flavors_into_database.py
Comment (schustmi) [inline src/zenml/zen_stores/migrations/versions/bea8a6ce3015_port_flavors_into_database.py:29]: workspace_id?

File: src/zenml/zen_stores/sql_zen_store.py
Comment (schustmi) [inline src/zenml/zen_stores/sql_zen_store.py:1627]: maybe `internal` or `builtin` instead of `non_custom`?

File: src/zenml/zen_stores/sql_zen_store.py
Comment (schustmi) [inline src/zenml/zen_stores/sql_zen_store.py:858]: I'm not sure how complicated it is to test this, but I think it would be quite critical to make sure this works as intended?

File: src/zenml/zen_stores/schemas/flavor_schemas.py
Comment (schustmi) [inline src/zenml/zen_stores/schemas/flavor_schemas.py:88]:
```suggestion
            logo_url=self.logo_url,
```

File: src/zenml/zen_stores/schemas/flavor_schemas.py
Comment (schustmi) [inline src/zenml/zen_stores/schemas/flavor_schemas.py:91]:
```suggestion
            flavor_docs_url=self.flavor_docs_url,
```

File: src/zenml/stack/flavor_registry.py
Comment (schustmi) [inline src/zenml/stack/flavor_registry.py:92]: `register_builtin_flavors` to keep it consistent with the property?

File: src/zenml/stack/flavor.py
Comment (schustmi) [inline src/zenml/stack/flavor.py:55]: Maybe we could make this optional, seems like quite a hassle for users getting started writing a custom integration? The dashboard could just display some placeholder image I guess?

File: src/zenml/stack/flavor.py
Comment (schustmi) [inline src/zenml/stack/flavor.py:38]: Again, this should probably be optional for custom integrations? And maybe shouldn't by default be implemented like this which is wrong for custom flavors?

I guess returning `None` here and then when registering it in the flavor registry, we could do
`flavor.flavor_docs_url or compute_default_docs_url()`?

Also, do we need the `flavor_` prefix or is `docs_url` fine as well?

File: scripts/verify_flavor_url_valid.py
Comment (schustmi) [inline scripts/verify_flavor_url_valid.py:35]: No :P

File: scripts/verify_flavor_url_valid.py
Comment (schustmi) [inline scripts/verify_flavor_url_valid.py:40]: Also no :P

Same for the rest of the commands

**Review by schustmi — CHANGES_REQUESTED** [review]
This looks great now with the sync, also really like the dev scripts/CLI!

Have some minor questions/renaming suggestions but nothing too wild

File: src/zenml/zen_stores/sql_zen_store.py
Comment (AlexejPenner) [reply src/zenml/zen_stores/sql_zen_store.py:1627]: built_in are our non-integration flavors, internal feels like it doesn't capture the integrations, basically this nukes everything that is not called 'custom' - feels like the most fitting name, even if it is a bit ugly

File: src/zenml/zen_stores/sql_zen_store.py
Comment (AlexejPenner) [reply src/zenml/zen_stores/sql_zen_store.py:858]: My worry as well, this also relies on the fact, that with every release we create an alembic revision

File: scripts/verify_flavor_url_valid.py
Comment (AlexejPenner) [reply scripts/verify_flavor_url_valid.py:35]: Can't I just copy-paste in peace? :D :P

Comment (AlexejPenner) [reply scripts/verify_flavor_url_valid.py:35]: Can't I just copy-paste in peace? :D :P

File: src/zenml/stack/flavor.py
Comment (AlexejPenner) [reply src/zenml/stack/flavor.py:55]: Yeah, I intended this to be optional, will make sure it will be.

File: src/zenml/stack/flavor.py
Comment (AlexejPenner) [reply src/zenml/stack/flavor.py:38]: docs_url is fine

Yeah you're right, we can generate this elsewhere, stil makes sense to have it as a filed though, so that users can link to their own docs.

File: src/zenml/zen_stores/sql_zen_store.py
Comment (AlexejPenner) [reply src/zenml/zen_stores/sql_zen_store.py:858]: I am not sure how to write these tests here, I think in general we need to write some testts, that our alembic migrations work as intended. It would then slot in well there

File: src/zenml/zen_stores/sql_zen_store.py
Comment (schustmi) [reply src/zenml/zen_stores/sql_zen_store.py:1638]: Good point, I was somehow confusing this with our actual builtin flavors. Not a big fan of `non_custom` but can't think of anything better either 👍

File: src/zenml/artifact_stores/local_artifact_store.py
Comment (schustmi) [reply src/zenml/artifact_stores/local_artifact_store.py:168]: should still be `s` instead of `S` right?

File: src/zenml/zen_stores/migrations/versions/bea8a6ce3015_port_flavors_into_database.py
Comment (schustmi) [inline src/zenml/zen_stores/migrations/versions/bea8a6ce3015_port_flavors_into_database.py:43]: docs_url also needs to be dropped I guess?

File: src/zenml/zen_stores/sql_zen_store.py
Comment (schustmi) [inline src/zenml/zen_stores/sql_zen_store.py:1642]: I think we had a similar problem with the filter model queries where we used `is` instead of sqlalchemy `.is_(...)`. This might also be a problem here?

File: src/zenml/zen_server/routers/workspaces_endpoints.py
Comment (schustmi) [inline src/zenml/zen_server/routers/workspaces_endpoints.py:487]: Aren't those still workspace scoped and need an endpoint here?

**Review by schustmi — CHANGES_REQUESTED** [review]

File: src/zenml/artifact_stores/local_artifact_store.py
Comment (AlexejPenner) [reply src/zenml/artifact_stores/local_artifact_store.py:168]: ah i fixed it in the s3 bucket but not here, done now

File: src/zenml/zen_server/routers/workspaces_endpoints.py
Comment (AlexejPenner) [reply src/zenml/zen_server/routers/workspaces_endpoints.py:487]: Flavors as a concept can't be workspace scoped, as such workspace_id needs to stay optional. However, for custom flavors we can still always ensure that they stay scoped.

File: src/zenml/zen_stores/sql_zen_store.py
Comment (AlexejPenner) [reply src/zenml/zen_stores/sql_zen_store.py:1642]: just to be safe I will switch to .is_()

Comment (AlexejPenner) [reply src/zenml/zen_stores/sql_zen_store.py]: So apparently you need to go with is_( when you are using the following API:

`getattr(table, "workspace_id").is_(None)`

**Pull request merged**
